DNM: Wip rgw d4n next

samarahu opened the pull request:
Updates for the single node cache implementation.

In addition to the testing outlined in the `wip-rgw-d4n-next` branch PR description, I executed the following ops

Non-versioned workflow:
1. `delete-object` with no specified version. This should completely delete the object from the cache, directory, and backend if not dirty.

Versioned buckets:
1. `delete-object` with no specified version. This should not delete anything and instead create a delete marker, which acts as the object's current version.
2. Simple `get`, which should return an `ENOENT` error because the object's latest version is a delete marker.
3. `delete-object` with the delete marker as the version. This should delete the delete marker and use the object's  previous version as the latest version.
4. Simple `get`, which should return the object's latest version.
5. Suspend the versioning for the bucket.
6. Simple `delete-object` with no specified version. This should not delete anything and instead create a delete marker with a value of "null".
7. Simple `get`, which should return an `ENOENT` error because the object's latest version is a delete marker.
8. `delete-object` with "null" as the version. This should delete the delete marker and use the object's previous version as the latest version.
9. Versioned upload of a new version for the same object, which should update the head object to contain the new version without overwriting the previous version by moving the previous version to the `prevVersion` field.
10. Versioned `delete-obj` with the object's first version. This should delete the object's first version while retaining the cache, directory, and policy entries for the latest version, as well as the head object.
11. `delete-object` with the latest version as a parameter. This should delete the latest version of the object from the cache, directory, and backend if not dirty. It should also remove the head object, completely deleting the object from all D4N structures.

mosayyebzadeh committed:
rgw/d4n: squashing all commits related to initial implementation
of write-back cache for non-multipart objects (writing objects to
cache and then to backend store in the cleaning process):

1. combining write cache with latest D4N code
2. cleaning the code
3. removing some bugs on bigger objects
4. updating iterate function to check the dirty flag
5. Updating write cache based on the pull request comments.
Read process needs to be updated based on write process. It needs to check where is the data and if it is dirty or clean.
If it is in the cache and dirty, we need to put D_ in the oid of the object before reading it from cache.
If it is clean, there is nothing to do.
6. updating flush functions and comments.

Signed-off-by: mosayyebzadeh <mosayyeb@bu.edu>

samarahu committed:
rgw/d4n: Add directory probing to read workflow

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: implementation of caching head in read workflow.

modifications in ReadOp::prepare() method of the d4n filter driver
to cache the head object.

modification in get_obj_attrs to read from cache or backend store.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: squashing all commits related to caching head in the
write-back workflow, modifying set_obj_attrs(), get_obj_attrs() and
delete_obj_attrs based on the cached head and modifying the cleaning
method to use a min-heap data structure for storing dirty objects only.

1. rgw/d4n: implementation for caching head object in write-back workflow.
2. rgw/d4n: modifications to get write back cache working after cleaning
process.
3. rgw/d4n: modifications for eviction of dirty blocks.
4. rgw/d4n: modifications include adding a heap of dirty objects
which has objects ordered by their creation time and the top
element of which is fetched in the cleaning method, processed
and deleted in a loop.
5. rgw/d4n: changing the format of cached blocks to
bucket_name_version_object_name_ofs_len, to avoid checks
for versioned and non-versioned objects.
6. rgw/d4n: modifications to set_obj_attrs(), modify_obj_attrs()
and delete_obj_attrs() to check if the head object exists in a cache,
else direct the calls to backend store.
7. rgw/d4n: handling version in case of bucket versioning being suspended
while writing the object.

Co-authored by: Samarah <samarah.uriarte@ibm.com>
Changed dynamic_cast to static_cast for D4NFilterObject
in D4NFilterWriter class constructors.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/d4n: squashing the following commits for miscellaneous fixes
to policy driver, redis driver, directory and filter driver.

1. d4n/directory: Use boost::split for simpler code
2. rgw: Lower log levels for failures in D4N and redis cache files
3. rgw: Add dpp and logs to directory, cache, and policy
4. rgw: Reduce Redis calls and fix workflow
5. qa/d4n: Remove D4N task and add S3 user creation to workunit driver script
6. d4n: Use Redis transactions to serialize consecutive requests for safe
data handling and faster completion
7. d4n/directory: Remove boost lexical_cast calls
8. rgw/d4n: Add return values to error logs
9. rgw/d4n: Change directory hostsList to use `unordered_set`
10. d4n/filter: Simplify logic for storing block in `handle_data`
11. rgw/policy: Properly delete `LFUDAEntry` instances
12. rgw/d4n: Add support for `dirty` block metadata, `check_bool` for consistent values, and fix directory updates in `cleanup` method

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n/test: squashing the following commits related to
D4N test failures.

1 d4n: Add `rename` method and unit test and update D4N unit tests
2. d4n/test: Fix linking issue

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: fix to correctly populate rgw_obj of GetObjectState in
RGWRados, in case ReadOp::prepare() reads the head object from
the cache.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: implementation of copyObject.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
d4n: Add filter `delete_obj` implementation and complete GET support

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: miscellaneous corrections/additions:
1. storing objects in directory using their oid, so that the version
is included.
2. making sure that the head block corresponds to latest
version in the block directory.
3. add a directory entry for head block for every version
in case of a versioned bucket.
4. Populating hostsList correctly for blocks and objects.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1207,28 +1213,126 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+
+    if ((ret = source->driver->get_cache_driver()->delete_data(dpp, prefix, y)) == 0) { // Sam: do we want del or delete_data here? 
+      if ((ret = source->driver->get_policy_driver()->get_cache_policy()->erase(dpp, policy_prefix, y))) {
+	if ((ret = blockDir->del(dpp, &block, y)) < 0) {
```
the deletion steps should also be reverse: 1. from directory 2. cache and metadata

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1207,28 +1213,126 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+
+      if ((ret = blockDir->get(dpp, &block, y)) < 0) {
+	ldpp_dout(dpp, 0) << "Failed to retrieve directory entry for: " << obj.objName << " blockid: " << fst << " block size: " << cur_len << ", ret=" << ret << dendl;
+	return next->delete_obj(dpp, y, flags);
```
at this point the loop must continue for other blocks in case they are in the cache. If the object is not dirty, then anyway next->delete() is invoked, which will take care of deleting all the blocks of the cache. And if a block/ object is dirty, then it would not have been evicted.

samarahu replied:
same question as below in regards to a block deletion failing

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1207,28 +1213,126 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+    block.size = 0;
+
+    // retrieve head object from directory to check for dirty flag
+    if ((ret = blockDir->get(dpp, &block, y)) < 0) { // Sam: can we somehow return the block from the check_head_exists_in_cache_get_oid method to avoid this redis call?
```
Like check_head_exists_in_cache_get_oid() returns attrs, it can return the `dirty` flag as well, if only the `dirty` flag is needed. also head_oid_in_cache, returns the oid of the head in cache, so there is no need for it to be derived in line 1259 etc.

samarahu replied:
You mean the `dirty` flag gets returned as part of the `head_oid_in_cache` string if the method returns `true`, correct?

pritha-srivastava replied:
well i had meant that the method return another bool param for a dirty flag, but what you have done with string parsing is also ok.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1207,28 +1213,126 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+      }
+
+      // check if block has been evicted
+      if (std::find(block.hostsList.begin(), block.hostsList.end(), dpp->get_cct()->_conf->rgw_local_cache_address) == block.hostsList.end() ) {
```
I think this step can be done away with

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1207,28 +1213,126 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+      }
 
-  return next->delete_obj(dpp, y, flags);
+      if (block.dirty == true)
```
At some point we must see if we can derive the `dirty` flag from the head block, as in the cleaning process, the head is marked clean only when all data blocks are marked clean, and Amin confirmed that all blocks in a write path are owned by the home node. But this can be marked as a TODO, as it will be clearer when distributed cache is in place.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1207,28 +1213,126 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+
+      std::string oid_in_cache = prefix + "_" + std::to_string(fst) + "_" + std::to_string(cur_len);
+
+      if ((ret = source->driver->get_cache_driver()->delete_data(dpp, oid_in_cache, y)) == 0) { // Sam: do we want del or delete_data here? 
```
same comment - remove entry from directory first.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1207,28 +1213,126 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+	if ((ret = source->driver->get_policy_driver()->get_cache_policy()->erase(dpp, policy_prefix + "_" + std::to_string(fst) + "_" + std::to_string(cur_len), y))) {
+	  if ((ret = blockDir->del(dpp, &block, y)) < 0) { 
+	    ldpp_dout(dpp, 0) << "Failed to delete directory entry for: " << obj.objName << " blockid: " << fst << " block size: " << cur_len << ", ret=" << ret << dendl;
+	    return next->delete_obj(dpp, y, flags);
```
Since the loop must continue for other blocks, this has to be removed : `next->delete_obj`

samarahu replied:
If the deletion of one block failed, shouldn't the entire operation fail? If it continues as normal and other blocks are properly deleted, the block that failed to be deleted is still taking up space. So instead of `next->delete_obj`, shouldn't I return an error?

pritha-srivastava replied:
Even if delete for a block fails and is taking up space, it will eventually be evicted. ok, so you can add a special check for -ENOENT and continue with the loop if this happens and return an error otherwise. Even though the block will eventually be evicted, the error must be communicated back to the end user. So instead of calling next->delete_obj, return the error.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1207,28 +1213,126 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+            }
+	} else {
+	  ldpp_dout(dpp, 0) << "Failed to delete policy entry for: " << obj.objName << " blockID: " << fst << " block size: " << cur_len << ", ret=" << ret << dendl;
+	  return next->delete_obj(dpp, y, flags);
```
This may not be needed , as dirty blocks do not exist in the backend store yet and deletion of a clean object is already taken care of in line 1332.

pritha-srivastava replied:
As discussed earlier, please add marking the object entry clean in the cleaning process, after the head block is marked clean.

samarahu replied:
> As discussed earlier, please add marking the object entry clean in the cleaning process, after the head block is marked clean.

Isn't this already handled in the `cleaning` method? I see the dirty flag being set to false for the head block and object near the bottom of the method. Please let me know if I'm misunderstanding what you meant.

pritha-srivastava replied:
I missed resetting the `dirty` flag for an object entry, I did it for the head object entry and all the data blocks.

samarahu committed:
rgw/d4n: Squash the following commits related to directory, eviction and filter driver logs:
1. d4n/directory: Remove `blockHosts` and change `objHosts` to store the hosts for the block
2. d4n/directory: Remove `dirtyBlock` metadata and use the cacheObj's `dirty` flag in block logic
3. rgw/d4n: This commit introduces the following changes:
  a. Fix `get_victim_block` method to correctly handle version
  b. Return `0` instead of error for non-error circumstances in `eviction`
  c. Switch `increase` to `decrease` for heap after localWeight update in
   `eviction`
  d. Update filter writer logs
4. d4n/filter: Make minor adjustments
5. test/d4n: Update `update_field` test calls

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava commented on src/rgw/driver/d4n/d4n_policy.cc:
```diff
@@ -591,16 +591,16 @@ void LFUDAPolicy::cleaning(const DoutPrefixProvider* dpp)
       block.blockID = 0;
       if (c_obj->have_instance()) {
-        dir->get(dpp, &block, null_yield);
+        blockDir->get(dpp, &block, null_yield);
```
why did this error happen in the first place? It built fine at my end and I tested things before pushing the last commit.

samarahu replied:
This wasn't an error. I added an `objDir` variable because I had to update the object in the `cleaning` process so I just wanted to differentiate both directory instances more clearly.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1272,30 +1279,117 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
-    ldpp_dout(dpp, 0) << "D4NFilterObject::" << __func__ << "(): CacheDriver del() method failed, ret=" << ret << dendl;
+    auto blockDir = source->driver->get_block_dir(); 
+    rgw::d4n::CacheBlock block;
+    block.cacheObj.objName = source->get_name();
```
this has been changed to get_oid()

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1272,30 +1279,117 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+    return next->delete_obj(dpp, y, flags);
+  } else {
+    rgw::d4n::CacheObj obj = rgw::d4n::CacheObj{ 
+			       .objName = source->get_name(),
```
it is source->get_oid() now

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1272,30 +1279,117 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+
+      if ((ret = blockDir->get(dpp, &block, y)) < 0) {
+	ldpp_dout(dpp, 0) << "Failed to retrieve directory entry for: " << obj.objName << " blockid: " << fst << " block size: " << cur_len << ", ret=" << ret << dendl;
+	return ret;
```
did we not discuss that deletion of other blocks will continue in case there is an error for a previous block

pritha-srivastava replied:
i remember now that we decided to return any error to the end user

samarahu replied:
Yes. If -ENOENT is returned, the rest of the blocks still get deleted, but other errors are returned.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1272,30 +1279,117 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
-  }
+    int ret = -1;
+    auto rgwObj = source->get_obj();
+    ret = source->get_obj_attrs(y, dpp, &rgwObj);
```
why is this call needed?

samarahu replied:
To populate the object's size since `source->get_obj_size()` is called later. If I don't call this method first, the value is 0.

pritha-srivastava replied:
You can check attrs returned by check_head_oid_exists_in_cache and get the object size using user.rgw.object_size. This way double calls to check_head_oid_exists_in_cache can be avoided

samarahu replied:
I removed this and I'm using the attr returned from `check_head_exists_in_cache_get_oid`

pritha-srivastava committed:
rgw/d4n: adding perfcounters for d4n cache hits, misses and evictions (LFUDA).

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -483,9 +492,14 @@ bool D4NFilterObject::check_head_exists_in_cache_get_oid(const DoutPrefixProvide
+  block.cacheObj.objName = this->get_name(); // account for blocks representing simple requests
   //if the block corresponding to head object does not exist in directory, implies it is not cached
-  if ((ret = blockDir->get(dpp, &block, y) == 0)) {
+  if (ret == 0 || (ret = blockDir->get(dpp, &block, y)) == 0) {
```
You have added a second get() call here, which will impact every read call, and I dont think this is acceptable. Can you justify why this is needed. The 'simple' read call that you are referring to here , is the one that doesnot have an instance as the input, and that will be taken care of by the get_oid() method, as it populates the object name correctly. Also even without a delete marker in place, if the head block is not found using get_oid(), then another request will be made using get_name(), and in case of an object not being present at all, we are simply sending two get() request to the redis server. Even for non-versioned objects. As I have pointed out before, you need to look into this. I had taken care of and tested the `simple` read call without the `delete_obj` changes and lets stick to that.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -392,7 +401,7 @@ void D4NFilterObject::set_obj_state_attrs(const DoutPrefixProvider* dpp, optiona
 {
   //versioned objects have instance set to versionId, and get_oid() returns oid containing instance, hence using id tag as version for non versioned objects only
-  if (! this->have_instance() && version.empty()) {
+  if (!this->have_instance() && version.empty()) {
```
as pointed out before, please avoid such formatting changes

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -483,9 +492,14 @@ bool D4NFilterObject::check_head_exists_in_cache_get_oid(const DoutPrefixProvide
   bool found_in_cache = true;
-  int ret = -1;
+  int ret = blockDir->get(dpp, &block, y);
+  block.cacheObj.objName = this->get_name(); // account for blocks representing simple requests
```
get_oid() will populate the object name correctly when no instance is provided. This will happen as part of the read_permission call which invokes get_obj_attrs(), which in turn calls this method. And later once set_instance is called and the prepare method again invokes this method, the get_oid() method will access the instance specific block entry. so add an entry for a delete marker instance, apart from the one that has the latest version. We can not afford to slow down the read path, as the delete might be called once, but `read` can be called multiple times.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -672,11 +708,50 @@ std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
+  block.size = 0;
+
+  int ret = -1;
+  if ((ret = blockDir->get(dpp, &block, y)) < 0) { 
```
this is one get() call here and then later get_obj_attrs_from_cache() invokes two more, even in cases where there are no delete markers. This needs to be modified.

pritha-srivastava committed:
rgw/d4n: squashing the following commits that fix s3 test
failures.

1. rgw/d4n: fixing logical errors for ranged requests in iterate(), aligning
start chunk/block to beginning of the chunk/block boundary.
2. cache head object in write path in case write-cache
is disabled - for maintaining cache consistensy
3. delete dirty blocks of older versions of a non-versioned
bucket as dirty blocks are not evicted, and older versions won't
be written to the disk also.
4. remove dirty objects from the dirty objects data structure,
in case the bucket it belongs to has been deleted.
5. rgw/d4n: fixing compilation issue after rebasing with main.
6. Handle multipart objects correctly in copy object
7. Fix more ranged request errors.
8. handling part_num correctly for multipart and non-multipart objects
9. handling obect name starting with '_'.
10. bug in range request for sending back data to the client, after reading from
backend store.
11. rgw/d4n: code to populate RGW_ATTR_CRYPT_PARTS attr
in d4n filter.
12. rgw/d4n: fix for 'if-match' and 'if-nomatch' for
put object and copy object.
13. rgw/cache: Account for slashes in object name
14. d4n/filter: Handle request headers during read if `next->prepare` isn't called
15. d4n/filter: Fix overwriting of object size and instance
16. rgw/d4n: Supply attrs to `calculate_version`
17. rgw/d4n: Handle ACL and etag in copy_object
18. rgw/filter: Ensure cksum values are maintained
19. rgw/d4n: Set parts_count correctly and check for bufferlist length in calculate_version
20. rgw/d4n: adding a case for iterate method, where last
part to read is smaller than chunk size.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>
Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -672,11 +705,36 @@ std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
+  ldpp_dout(dpp, 0) << "D4NFilterObject::" << __func__ << "()" << dendl;
+  auto blockDir = source->driver->get_block_dir(); 
+  rgw::d4n::CacheBlock block;
+  block.cacheObj.objName = source->get_name(); // for simple ops, the name will be equivalent to the oid, so we want to use the name to retrieve the block
```
This extra logic to get the block and then check whether there is a delete marker is not needed. get_obj_attrs_from_cache() will return an -ENOENT in case of a delete marker, correct?

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -677,9 +703,14 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
   bool is_latest_version = true;
   if (source->have_instance()) {
-    is_latest_version = false;
+    is_latest_version = false; // what about the case where the version supplied is the latest version? -Sam
```
@pritha-srivastava You set `is_latest_version` to `false` if an instance is found but couldn't the instance also be the latest version if that particular version id is supplied to the request?

pritha-srivastava replied:
Yes, thats ok, in that case, the block containing the head version will not be set to the latest version. Otherwise the instance block for that version will be present in the cache, or will be cached, and can be used to get back the head information. In case of the write-back cache, every new object written will be the latest version, and hence the head block will always have the latest version. Only in a read only workflow, if the user sends a request with the version id which is also the latest version, then the head object will be fetched from the backend store, and generally the latest version is read without supplying the version id. I do not want to add an extra get() call just to serve this scenario because this will impact all read workflows. I am exploring alternate ways to fix this corner case. Please don't make any changes to the readop::prepare() for this. 
This will not impact the delete_obj workflow if you have added code to correctly store a block instance for every delete marker also - the method `check_head_exists_in_cache_get_oid` will correctly fetch the block/ block instance based on whether the version is supplied, using get_oid(). If the block is not cached, the correct error will be returned from the backend store

samarahu replied:
Ok, I will remove my comment since you are already looking into this. I haven't made any changes to `prepare` in regards to this corner case and I've updated the `delete_obj` logic to store a block instance to represent delete markers so the workflows should work as intended.

pritha-srivastava replied:
All workflows need to be tested and confirmed before I cherry-pick the delete_obj commit, including creation of delete markers and deletion of delete markers and for non-versioned buckets.

samarahu replied:
Yes, I've confirmed these workflows complete successfully and display the correct values in the directory.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1432,41 +1464,175 @@ int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl
+      } 
+      
+      deleteBlock = block;
+      deleteBlock.cacheObj.objName = "_:" + deleteBlock.version + "_" + deleteBlock.cacheObj.objName;
```
can't source->get_oid() be used here?

samarahu replied:
Not in this case, because this is for the situation where there is no version provided to the request, which means the oid would not contain the version.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1432,41 +1464,175 @@ int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl
+	return ret;
+      }
+      
+      if (block.cacheObj.objName != source->get_name()) {
```
what is being done here? please add one line comment in sections of code describing what is being done

samarahu replied:
This is to account for the case where the delete request is versioned and a non-versioned head object still exists from previous simple requests. This won't apply to workflows without versioned buckets because the oid and name will match in those cases.

pritha-srivastava replied:
can you describe the test steps which will lead to this scenario? what is a non-versioned head object? is it the one that contains the latest version?

samarahu replied:
The non-versioned head object is the object that contains the latest version but doesn't contain the version in the key. During a PUT in a versioned bucket, the non-versioned head object is created and set by the `set_head_obj_dir_entry` method using the object name (so the block does not use the version in the key). During a versioned delete on this same object, the head object will be left over. However, now that I'm considering this, I think the head object should be left over because other versions can still exist for that object and the head object should still exist to represent this state. I would have to update the head object to contain the correct current version after the supplied version is deleted and delete the head object if there are no more versions of the object left.

pritha-srivastava commented on src/rgw/driver/d4n/d4n_directory.cc:
```diff
@@ -178,7 +178,7 @@ int ObjectDirectory::get(const DoutPrefixProvider* dpp, CacheObj* object, option
     object->objName = std::get<0>(resp).value()[0];
     object->bucketName = std::get<0>(resp).value()[1];
     object->creationTime = std::get<0>(resp).value()[2];
-    object->dirty = std::stoi(std::get<0>(resp).value()[3]);
```
You must modify the commit message to remove the word 'simple' from the commit message, as get requests for all scenarios work even when an instance is supplied.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1432,41 +1464,177 @@ int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl
+      deleteBlock.cacheObj.objName = "_:" + deleteBlock.version + "_" + deleteBlock.cacheObj.objName; // since the request has no instance,
+												      // the oid does not contain the version
+
+      if ((ret = blockDir->set(dpp, &deleteBlock, y)) == 0) {
```
ok, I get it now, this line sets the instance head block for versioned buckets, correct?

samarahu replied:
`deleteBlock` represents the block that contains the delete marker in the key. This way, if another request comes in to delete the delete marker, `deleteBlock` will be easily retrieved and deleted.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1432,41 +1464,177 @@ int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl
+	ldpp_dout(dpp, 0) << "Failed to delete delete marker block in block directory for: " << block.cacheObj.objName << ", ret=" << ret << dendl;
+	return ret;
+      }
+    }
```
You need to call eraseObj to erase the object from the map/object heap (this only applies to dirty objects).
There is also an entry for the object in objDir, which needs to be deleted (regardless of whether the object is dirty or not).

samarahu replied:
`eraseObj` is called in the `cleaning` method; can you explain why it's called there rather than being updated to not have a dirty flag? The `o_entries_map` would no longer contain this object entry and attempting to delete it during a `delete_obj` call would result in a failure.

pritha-srivastava replied:
like I said in my previous comment o_entries_map is used to maintain only dirty objects, hence cleaning calls eraseObj(); there is a call to objDir->update_field, to update the dirty flag correctly. In case delete obj is called for a dirty object, eraseObj() needs to be called for the reason I explained. If it is still dirty, erasing it from o_entries_map will not result in a failure.

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1609,12 +1667,22 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
       deleteBlock = block;
       block.cacheObj.objName = source->get_name();
 
+      if (block.version == "null") 
```
For some reason, I observed that requests with version-ids set to `null` have an oid that is equivalent to the object name, which is why I handle `null` version blocks this way.

pritha-srivastava replied:
ok

samarahu committed:
d4n/filter: Implement `get_bucket` and `load_bucket`

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -832,6 +832,27 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
       ldpp_dout(dpp, 0) << "D4NFilterObject::" << __func__ << "(): failed to cache head object during eviction, ret=" << ret << dendl;
     }
   } else {
+    bufferlist etag_bl;
```
this I think should only be done, after get_obj_attrs_from_cache(), otherwise the call to backend store should take care of it. check this RGWRados::Object::Read::prepare(), the backend store does take care of these preconditions.

samarahu replied:
Yes, the call to `get_obj_attrs_from_cache` is at the beginning of this if-else statement so once this `else` statement is reached, that method would have already been executed

pritha-srivastava replied:
ok this piece of code looks correct

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1662,8 +1688,14 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
-	  return ret;
+          if (ret == -ENOENT) {
+	    ldpp_dout(dpp, 0) << "Directory entry for: " << source->get_name() << " blockid: " << fst << " block size: " << cur_len << " does not exist; continuing" << dendl;
+            fst += g_conf()->rgw_get_obj_max_req_size; // does this handle range requests? -Sam
```
I don't think this is correct, I had explained in my comment in your previous PR how to correctly increment fst

samarahu replied:
May you please point me to that information? I attempted to search for it but wasn't able to find it

pritha-srivastava replied:
Check this: https://github.com/samarahu/ceph/commit/d75a3903e04cabf8059477bdee2fa7fea7f7961c#diff-5cf52d0baced08513b69bbb3cf4a29682289fce58451b9567c6d2e86ff080bceR1668

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -575,6 +574,9 @@ bool D4NFilterObject::check_head_exists_in_cache_get_oid(const DoutPrefixProvide
     }
+    
+    // get_attrs overwrite version, so set below
+    this->set_object_version(version);
```
Can you explain how get_attrs() overwrites the version? and which get_attrs()? cachedriver->get_attrs() does not deal with the object version at all

samarahu replied:
Removing this change because it's unnecessary.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1634,7 +1636,7 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
 
     if (block.deleteMarker == false) { // provided version is not a delete marker and contains data
-      if (source->get_bucket()->versioned()) { 
+      if (source->get_bucket()->versioned() && block.cacheObj.objName != source->get_name()) { 
```
What is source->get_name() and under what conditions will it not be the same as block.cacheObj.objName? Please elaborate so that I can understand what is going on here.

samarahu replied:
During the `null` delete workflow, I found that the object name doesn't have the version (so it isn't versioned) but it still satisfies the condition on this line, which leads to a BlockDirectory::del on the versioned head object. Later on in the method, the non-versioned head object is also deleted but since the non-versioned head object already gets incorrectly deleted on line 1640, there is an error. As a result, I added an extra condition that ensures the head object is versioned: if the block's cacheObj name does not match the simple and unversioned object name (`source->get_name()`), then the block's cacheObj must be versioned.

pritha-srivastava replied:
Ok

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -832,6 +832,47 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
     }
   } else {
+    bufferlist etag_bl;
+    if (get_attr(dpp, RGW_ATTR_ETAG, etag_bl, y) < 0) {
```
can you separate out changes related to `Handle request headers during read if `next->prepare` isn't called` from changes related to delete_obj. I believe you are still working on delete_obj so I can cherry-pick this change and this will help solve some of the errors that I am seeing in my testing exercise.

pritha-srivastava committed:
rgw/d4n: removing code that writes object entry to the
object directory in redis

Objects were being written to the directory for bucket listing
but another data structure needs to be used that allows for
ordered listing indexed by a bucket name/ id, hence removing
the entries that do not serve this purpose (they also add to the
number of calls to the redis server).

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: implementation of list objects for a bucket
(ordered listing) and list-object-versions.

Changes include the addition of an ordered set for
each bucket that stores the object, addition of etag
and object size to CacheObj structure needed for listing
objects, and addition of a test case to see the usage of
pipelining in Redis.  Objects are fetched from the cache
and the backend store and are then merged, before returning
the final list.

Co-authored-by: sinashan <sinatak1373@live.com>
Added code for bucket listing using ObjectDirectory(this
has been removed)

Co-authored-by: Samarah <samarah.uriarte@ibm.com>
Fixed log levels in BucketDirectory and ObjectDirectory
methods. Added enum for field retrieval in Directory get
methods.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1028,8 +1028,11 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
     bufferlist etag_bl;
-    if (get_attr(dpp, RGW_ATTR_ETAG, etag_bl, y) < 0) {
-      return -EINVAL;
+    this->source->load_obj_state(dpp, y);
```
please address this as discussed in the chat.

pritha-srivastava committed:
rgw/d4n: miscellaneous updates:

1. use bucket id instead of name in the cache block key
as this will be helpful to differentiate between objects
belonging to buckets that have been deleted and recreated.
2. add helper methods to derive prefix and key for cache blocks.
3. comments and code cleanup, renaming of policy methods
related to adding and erasing dirty objects to and from
in-memory data structures.
4. rgw/d4n: reading in chunks of rgw_max_chunk_size instead
of rgw_obj_max_req_size since data is being written in rgw_max_chunk_size
chunks in the write back cache.
5. rgw/d4n: correctly set the version for objects
with 'null' versionId in case of a write workflow.
The version stored in the block directory was 'null',
but that does not differentiate two different versions
of objects with versionId set to 'null', hence differentiating
between objects with 'null' versionId by deriving and storing
a non-null version in the block directory, and based on this
version the name of the object in the cache is derived.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -633,7 +633,11 @@ int D4NFilterObject::set_data_block_dir_entries(const DoutPrefixProvider* dpp, o
     off_t cur_len = cur_size - fst;
     block.cacheObj.bucketName = this->get_bucket()->get_name();
-    block.cacheObj.objName = this->get_key().get_oid();
+    if (this->get_bucket()->versioned() && !this->get_bucket()->versioning_enabled()) {
```
why is this check needed? the oid must be used to locate the block, else how will you differentiate between an instance when versioning gets suspended and when it is not enabled at all. you need to ensure that version is set to 'null'. doesn't is_instance() return null for versioning suspended?

pritha-srivastava replied:
Even here, the version is calculated in writer::prepare() method and used in process() and complete()

pritha-srivastava replied:
if the problem is that ReadOp::prepare method does not populate version for the case when versioning is suspended, then in calculate_version method, you need to add the check that if versioning is suspended and then set the instance to 'null' after this line: https://github.com/ceph/ceph/pull/56336/files#diff-5cf52d0baced08513b69bbb3cf4a29682289fce58451b9567c6d2e86ff080bceR516. Something like the following:

```
version = "null";
object->set_instance(version);
```
And then do away with all the other checks.

samarahu replied:
get_instance() returns an empty string when versioning is suspended. For suspended versioning, the block directory will not have "null" in the block name but the block will store "null" as the version. I'll make the suggested change to `calculate_version` and update the rest of the logic

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1197,7 +1201,11 @@ int D4NFilterObject::D4NFilterReadOp::iterate(const DoutPrefixProvider* dpp, int
   std::string version = source->get_object_version();
   std::string prefix;
 
-  prefix = source->get_bucket()->get_name() + "_" + version + "_" + source->get_name();
```
instead of doing this, you need to reset the version correctly.

pritha-srivastava replied:
if versioning is suspended, the version calculated must have automatically been set to 'null', if it isn't you need to fix that. the idea here is to calculate the version in prepare() and then use it in iterate() method

samarahu replied:
Got it, I'll make sure the version is correctly being calculated in `prepare`

pritha-srivastava committed:
rgw/d4n: squashing the following commits for restoring
in-memory data for LFUDA policy and dirty objects from
cache on disk

1. rgw/d4n: restore in memory data structure for dirty
objects using the xattrs in the head block of an object
in the cache.
2. rgw/d4n: replacing string cache attr names with constexpr char.
3. rgw/d4n: restore LFUDA policy data from cache on disk.
4. rgw/d4n: correcting the key used while updating the LFUDA data structure
for the head object block in the read path.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: handling special characters in object names such that
parsing file names while restoring data from cache does not result
in errors.

1. Introduce a new delimiter # instead of _ for concatening bucket_id,
objectname etc to create a key, because "_" is a valid char in object names
and is not encoded and results in incorrect values while parsing filenames
for restorin data.
2. url encode the bucket id, version and object name before all are concatenated
to form a key
3. url_decode the bucket id, version and object name after parsing the filename
from cache while restoring data back.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -612,6 +612,42 @@ int D4NFilterObject::set_head_obj_dir_entry(const DoutPrefixProvider* dpp, optio
+      .cacheObj = null_object,
+      .blockID = 0,
+      .version = this->get_object_version(),
+      .prevVersion = block.prevVersion,
```
is there going to be a prevVersion here, or are you just doing it for the sake of completeness?

pritha-srivastava replied:
I mean is there a need to do a get and set here for this block? can we not safely assume that null_block.prevVersion = {null_block.version, null_block.deleteMarker}; always?

pritha-srivastava replied:
And you haven't set the deleteMarker here, is that intentional and if so why?

samarahu replied:
It was just for the sake of completion but I agree with your point, so I will remove it. In this case, there is also no possibility for a delete marker to occur, correct?

pritha-srivastava replied:
Correct.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1809,7 +1809,7 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
 	  tempBlock.cacheObj.objName = "_:null_" + block.cacheObj.objName;
 	  ret = blockDir->get(dpp, &tempBlock, y); // check if null delete marker block already exists; if so, delete
-	  if (ret == 0) {
+	  if (ret == 0 && tempBlock.deleteMarker) {
```
I do not have the latest changes of delete_obj so this change didnt make it in the cherry-picked commit, You can make a note of this and add it to your final delete_obj commit.

samarahu committed:
rgw/d4n: squashing commits for supporting delete object
for unversioned buckets.

1. d4n/directory: Update logging
2. d4n/filter: Update `set_head_obj_dir_entry` to handle null versioning and update `prevVersion` for versioned head objects
3. rgw/d4n: Simplify code to only support unversioned bucket deletes

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1816,48 +1812,85 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+      auto tempBlock = block;
+
+      if (!source->get_bucket()->versioning_enabled()) { // if versioning is suspended
+        if (block.version == source->get_object_version()) {
```
check_head_exists_in_cache_get_oid() method sets the version for the object using the version in the block retrieved, so why is this check needed? Also can you add detailed comments for every case here

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -612,6 +612,39 @@ int D4NFilterObject::set_head_obj_dir_entry(const DoutPrefixProvider* dpp, optio
+      .size = 0,
+    };
+
+    ret = blockDir->get(dpp, &null_block, y);
```
Is there really a need for a 'get' and then 'set', can we just not overwrite the old values, because the most recent one is valid, we do not need to update the hostsList etc also, as this block found in the cache belonging to the local node is the only valid "null" block, so I believe its safe to just overwrite the existing values.

samarahu replied:
I agree, but can this not occur for the versioned head object logic above as well? Either the block is set from scratch or only certain fields are updated with new values, which will happen regardless.

pritha-srivastava replied:
Well for the versioned head object you had made these changes while storing the prevVersion and deleteMarker fields. Now that I look at this line of code: version_block.prevVersion = {version_block.version, version_block.deleteMarker}; it appears incorrect to me, shouldn't prevVersion be set to block.version before setting the new version for block? I am not sure whether you had any other reason in mind for making a get and set call for a versioned head block, but it appears that you only need to do a set and also correct the prevVersion field value.

pritha-srivastava replied:
We do need the version chaining in the block though.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -612,6 +612,39 @@ int D4NFilterObject::set_head_obj_dir_entry(const DoutPrefixProvider* dpp, optio
+
+      ret = blockDir->set(dpp, &null_block, y);
+      if (ret < 0) {
+	ldpp_dout(dpp, 10) << "D4NFilterObject::" << __func__ << "(): BlockDirectory set method failed for head object with ret: " << ret << dendl;
```
Modify the comment to indicate a null head object

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1816,48 +1812,85 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
   std::string head_oid_in_cache;
   rgw::d4n::CacheBlock block;
   if (!source->check_head_exists_in_cache_get_oid(dpp, head_oid_in_cache, attrs, block, y) && !block.deleteMarker) {
```
you must add a comment that check_head_exists_in_cache_get_oid will return false if the block exists in cache but is a delete marker, which implies that this condition is satisfied only when the head block is not in the cache.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1816,48 +1812,85 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
     bool objDirty = false;
     auto blockDir = source->driver->get_block_dir(); 
     std::string version, policy_prefix;
+    bool nullDelete = false;
```
what if the object is dirty? Are these check common for dirty and clean objects?

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1816,48 +1812,85 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+
+	ldpp_dout(dpp, 10) << "D4NFilterObject::" << __func__ << "(): retrieving delete marker; calling next->delete_obj" << dendl;
+	next->params = params;
+	ret = next->delete_obj(dpp, y, flags);
```
next->delete_obj is being called without checking if the object is dirty or not.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1868,7 +1901,16 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+	    }
+	  } // else, only delete the data blocks since the null head object directory entry was updated to a delete marker
+	  
+	  ldpp_dout(dpp, 10) << "D4NFilterObject::" << __func__ << "(): deleting versioned head object data block for " << head_oid_in_cache << dendl;
```
again, below cache blocks are being deleted without a check whether they are dirty or not. We had decided that non-dirty/ clean blocks will be taken care of during eviction

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1878,31 +1920,53 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+		return ret;
+	      }
+
+	      if ((ret = source->driver->get_obj_dir()->del(dpp, &headObj.cacheObj, y)) < 0 && ret != -ENOENT) {
```
This doesn't exist anymore, in one of my latest commits I have removed the objects being stored in the object directory, as we need to come up with a better data structure to store objects indexed by bucket name. So you can probably leave a place holder here by adding a comment

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1946,8 +2005,17 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
 	if ((ret = blockDir->get(dpp, &block, y)) < 0) {
-	  ldpp_dout(dpp, 0) << "Failed to retrieve directory entry for: " << source->get_name() << " blockid: " << fst << " block size: " << cur_len << ", ret=" << ret << dendl;
-	  return ret;
+          if (ret == -ENOENT) {
```
did we not decide that for non-dirty objects, the deletion will be taken care of by eviction. We only need to remove the block directory entries so that the object is not found in the cache anymore.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -1816,48 +1812,85 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
   if (!source->check_head_exists_in_cache_get_oid(dpp, head_oid_in_cache, attrs, block, y) && !block.deleteMarker) {
-    ldpp_dout(dpp, 0) << "D4NFilterObject::D4NFilterDeleteOp::" << __func__ << "(): calling next delete_obj" << dendl;
-    return next->delete_obj(dpp, y, flags);
+    ldpp_dout(dpp, 10) << "D4NFilterObject::" << __func__ << "(): head object not found; calling next->delete_obj" << dendl;
```
I have a comment to organize the overall structure of the code:
1. find the object in the cache, if not present, call next->delete.
2. If found in cache, perform the following steps, common ones for dirty and clean objects are 3&4.
3. Insert a delete marker or delete a particular version of the head object based on the input, which includes a delete marker also. Perform all block directory manipulations like updating the previous version, deleteMarker etc for the head object.
4. Delete block directory entries for the data blocks.
5. If the object is dirty, delete data blocks. if not, call next->delete to delete the object from the backend store.

You could also combine step 1& 5 for non-dirty objects.

samarahu replied:
I believe the method currently follows this general structure but I reorganized the code slightly so that delete markers are dealt with first

pritha-srivastava committed:
rgw/cache: implementation for organizing data in ssd cache
into bucket directory, object directory and then versions
representing file names.

This commit for extracts bucket id, object name, version,
offset, and length from input key and organizes data into
directories - with bucket id, object name being used to
create directories and filenames within them using version
and offset and length.

Data can be restored back to in memory LFUDA and dirty object
data structure once rgw is restarted.

Updating unit tests.
test/rgw: Correct RedisDriver keys and add checks for cache blocks after
write

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>
Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/directory: support for ordered set in both ObjectDirectory
and BlockDirectory.

Adds methods to add, list a range and remove members from a
redis ordered set ordered by score passed in. Also adds unit
test cases to test the methods.

Samarah has fixed the arguments passed in to zrevrange in
ZAddZRemYield and ZAddZRevRangeYield tests.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: support for dirty objects in delete object method.

1. ordered set to maintain versions of a dirty object
2. creation of delete marker in case of a simple delete request
3. deletion of a specific version from the ordered set
4. cleaning method deletes from ordered set for dirty objects
5. use of redis atomicity constructs wherever needed

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: squashing following commits that add support for
WATCH, EXEC, INCR, MULTI and DISCARD redis commands.

1. d4n/directory: support for watch, exec and incr methods in Block
and Object Directory.
2. d4n/directory: support for redis MULTI and DISCARD command,
adding a test case to exercise usage of MULTI.

Co-authored-by: Samarah <samarah.uriarte@ibm.com>
Adjusted log levels for directory methods added in this commit.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: squashing commits related to handling of
clean (non-dirty) objects for delete_obj method.

1. rgw/d4n: miscellaneous fixes that include modifying
set_head_obj_dir_entry() and delete_obj method to handle
clean objects only.
2. rgw/d4n: changes to design when write cache is turned off:
a. do not store a hash entry for latest version for versioned
buckets, the latest version will always be fetched from backend store,
to maintain correct order of versions during a delete object request.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/d4n: squashing following commits related
to lazy deletion - delete dirty objects data
blocks in cleaning method and non-dirty objects
data blocks in eviction method

1. rgw/d4n: Implement lazy deletion
2. rgw/d4n: cleaning method now supports delete markers
also (both versioned and null). The delete markers
are written correctly to the backend store.
Also modifying the description of rgw_d4n_cache_cleaning_interval
to explicitly state that the duration is in seconds.
3. rgw/d4n: do not call invalidate_dirty_object in case
of a simple delete request for dirty objects belonging
to a versioned bucket.
In this case, a delete marker needs to be created instead of
invalidating/deleting an object.
4. rgw/d4n: Update lazy deletion in policy

Co-authored-by: Pritha Srivastava <prsrivas@redhat.com>
Added code for creation of delete marker for a simple
delete request.

Signed-off-by: Samarah <samarah.uriarte@ibm.com>
Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

Commit 2c1adbbd2b363cd5e02fd7da0cb496ca6a93aa77

d4n/policy.cc: Use ceph::split

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
rgw/cache: Remove `del` method

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: squashing following commits to address concurrency in
filter driver and ssd backed cache.

1. rgw/d4n: addressing concurrency issues by adding a refcount
to each block.
Blocks with positive refcounts are pinned (similar to dirty blocks)
and not eligible for eviction or deletion. Updating unit tests also.
2. rgw/cache: addressing concurrency issues while directories creation,
deletion, updating xattrs, get_attr and put.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/d4n: Remove dirty prefix and make xattr instead

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n/directory: Add more error handling to update_field

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

mkogan1 committed:
rgw/d4n: fix valgrind Invalid Memory Read & Write

Signed-off-by: Mark Kogan <mkogan@redhat.com>

mkogan1 committed:
rgw/d4n: fix output of binary data to text log and logs cleanup

the RGW_ATTR_ACL attr logging requires special handling

Signed-off-by: Mark Kogan <mkogan@redhat.com>

pritha-srivastava committed:
rgw/d4n: removing calls to redis MULTI, DISCARD, WATCH,
UNWATCH and EXEC since other options for Redis transactions
are being explored.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

mkogan1 committed:
rgw/d4n: write to the cache with O_SYNC flag by default

add conf to allow modifying cache write 'man 2 open' flags
with O_SYNC enabled by default.

performance impact measurment -
workload is PUT of 100K 5KB objects

- with O_SYNC flag: 3890 OP/s,
  iostat IO util% range while workload is running: ~45%-49%

- without O_SYNC flag: 4511 OP/s,
  iostat IO util% range while workload is running: ~10%-25%

Signed-off-by: Mark Kogan <mkogan@redhat.com>

pritha-srivastava committed:
rgw/d4n: delete older versions of an object
during an overwrite for objects belonging to
an unversioned bucket and version suspended
bucket in Writer::complete() method.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/d4n: making LFUDAPolicy destructor lock free to prevent
deadlocks during rgw shutdown.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>
Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
test/d4n: Add D4N filter test program

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: implementation of load_obj_state() and exists()
for D4NFilterObject.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/d4n: squashing commits related to changing
creationTime type from time_t to double.

1. rgw/d4n: Change score calculation
2. rgw/d4n: Update creationTime calculation to include decimal value
3. rgw/d4n: changes for converting string value to
time_point, when creationTime type changed to double.

Co-authored-by: Pritha Srivastava <prsrivas@redhat.com>
Updated bucket::list to convert from string to time_point

Signed-off-by: Samarah <samarah.uriarte@ibm.com>
Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava reviewed (commented):
I have added this commit to your previous commit related to removing del from redis driver and squashed the two

pritha-srivastava commented on src/rgw/driver/d4n/d4n_policy.cc:
```diff
@@ -887,7 +887,7 @@ void LFUDAPolicy::cleaning(const DoutPrefixProvider* dpp)
 	    } //end-if (block.version == entry->version)
 	  } //end - else if op_ret == 0
-	  ldpp_dout(dpp, 10) << "D4NFilterObject::" << __func__ << "(): Removing object name: "<< c_obj->get_name() << " score: " << std::setprecision(std::numeric_limits<double>::max_digits10) << e->creationTime << " from ordered set" << dendl;
+	  ldpp_dout(dpp, 10) << "D4NFilterObject::" << __func__ << "(): Removing object name: "<< c_obj->get_name() << " score: " << std::setprecision(std::numeric_limits<time_t>::max_digits10) << e->creationTime << " from ordered set" << dendl;
```
@samarahu : why is this needed? didn't we change time format to double for greater precision everywhere?

samarahu replied:
This is in regards to one of Matt's comments about using a time type instead of `double` but I understand your point. Please feel free to discard this change; I will share this with Matt

pritha-srivastava commented on src/rgw/driver/d4n/d4n_directory.h:
```diff
@@ -64,7 +64,7 @@ class ObjectDirectory: public Directory {
     int copy(const DoutPrefixProvider* dpp, CacheObj* object, const std::string& copyName, const std::string& copyBucketName, optional_yield y);
     int del(const DoutPrefixProvider* dpp, CacheObj* object, optional_yield y);
-    int update_field(const DoutPrefixProvider* dpp, CacheObj* object, std::string field, std::string value, optional_yield y);
+    int update_field(const DoutPrefixProvider* dpp, CacheObj* object, const std::string& field, std::string& value, optional_yield y);
```
did you check whether this compiles fine, it gives me errors

samarahu replied:
What is the error? It compiles on my end so that is strange

pritha-srivastava replied:
there is a follow up commit that fixes the error

samarahu replied:
Apologies, I meant to squash all commits related to this change into one.

adamemerson committed:
rgw/sal: Add default shutdown method to filters

It simply calls `next->shutdown()`.

Fixes: https://tracker.ceph.com/issues/70771
Signed-off-by: Adam C. Emerson <aemerson@redhat.com>

adamemerson committed:
rgw/d4n: Shutdown Redis connection in `shutdown()`

Since we changed the shutdown logic to let Asio tasks finish, make
sure we finish.

Fixes: https://tracker.ceph.com/issues/70771
Signed-off-by: Adam C. Emerson <aemerson@redhat.com>

adamemerson committed:
rgw/d4n: Use `unique_ptr` rather than bare new/delete

Signed-off-by: Adam C. Emerson <aemerson@redhat.com>

samarahu committed:
qa/d4n: Update D4N suite

Signed-off-by: Samarah <samarah.uriarte@ibm.com>
